packages/api/src/car.js

```diff
@@ -19,44 +23,90 @@ export async function carHead(request, env, ctx) {
-    // if (request.headers.get('Accept-Encoding').match('gzip')) {
-    //   headers['Content-Encoding'] = 'gzip'
-    // }
-    res = new Response(res.body, { ...res, headers })
-    ctx.waitUntil(cache.put(request, res.clone()))
+
+  if (res) {
+    return res
+  }
+
+  const {
+    params: { cid },
+  } = request
+  // gateway does not support `carversion` yet.
+  // using it now means we can skip the cache if it is supported in the future
+  const url = new URL(`/api/v0/dag/export?arg=${cid}&carversion=1`, GATEWAY)
+  console.log(url.toString())
+  res = await fetch(url, { method: 'POST' })
+  if (!res.ok) {
+    // bail early. dont cache errors.
+    return res
+  }
+  // TODO: these headers should be upstreamed to ipfs impls
+  // without the content-disposition, firefox describes them as DMS files.
+  const headers = {
+    'Cache-Control': 'public, max-age=31536000',  // max max-age is 1 year
+    'Content-Type': 'application/car',
+    'Content-Disposition': `attachment; filename="${cid}.car"`,
   }
+  // // compress if asked for? is it worth it?
+  // if (request.headers.get('Accept-Encoding').match('gzip')) {
+  //   headers['Content-Encoding'] = 'gzip'
+  // }
+  res = new Response(res.body, { ...res, headers })
+  ctx.waitUntil(cache.put(request, res.clone()))
+
   return res
 }
 
 /**
  * @param {import('./user').AuthenticatedRequest} request
+ * @param {import('./env').Env} env
  */
-export async function carPost(request, env, ctx) {
-  const { user, authToken } = request.auth
-  console.log('User ID', user._id)
-  console.log('Auth Token ID', authToken && authToken._id)
-  return new JSONResponse({})
+export async function carPost(request, env) {
+  const { authToken } = request.auth
+
+  const { headers } = request
+  const contentType = headers.get('content-type')
+
+  if (!contentType.includes('application/car')) {
+    return new Response(`${request.method} /car did not receive a car content type`, { status: 400 })
+  }
+
+  let name = headers.get('name')
+  if (!name || typeof name !== 'string') {
+    // throw new Error('invalid name') // Is it required?
+    name = 'testname'
+  }
+
+  const blob = await request.blob()
+
+  // Ensure car blob.type is set; it is used by the cluster client to set the foramt=car flag on the /add call.
+  const content = blob.slice(0, blob.size, 'application/car')
+
+  const { cid } = await cluster.add(content, {
+    // When >2.5MB, use local add, because waiting for blocks to be sent to
+    // other cluster nodes can take a long time. Replication to other nodes
+    // will be done async by bitswap instead.
+    local: blob.size > LOCAL_ADD_THRESHOLD,
+  })
+
+  // Store in DB
+  await env.db.query(gql`
+    mutation importCar($data: importCarInput!) {
+      importCar(data: $data) {
+        _id
+      }
+    }
+  `, {
+    data: { 
+      authToken,
+      cid,
+      name
+      // dagSize: undefined // TODO: should we default to chunk car behavior?
```

Comment:
We kinda want this to be `null` if partial DAG or `blob.size` otherwise...

Not sure how to do this without reading through the CAR to see if it has any missing blocks...which I guess we _could_ do.

Reply:
If we did then we could default the name to `hamster.jpg` when the user uploads a CAR file of a directory with 1 file.

Reply:
Then we will get the pin status from cluster, and I think we can get the size from it if single upload?

So, I was thinking that to be a better approach than identifying if this is a partial car. wdyt?

Reply:
I don't understand...?

Reply:
The next step on POST car is to get the pin status from cluster. If the cluster receives the entire CAR in this request, it will already have the size for us in its reponse.
But, I was now thinking and it might go to the request queue

Reply:
> If we did then we could default the name to hamster.jpg when the user uploads a CAR file of a directory with 1 file.

humm, so you mean really unpack the entire CAR file?

Reply:
Given we know in the client if it is chunked, I think a simple and more efficient approach would be to add a Header `X-Chunked`. If it exists we do not care about the bytes returned by `cluster.add`, otherwise we set them as size in DB.
What do you think?

Reply:
after syncing, we will start by having the size when pinned, for simplicity and guarantee data correction.

Reply:
It's a good idea, but I don't think we can rely on it being sent, or it being sent in the correct situations. We're in charge of our JS client but it's a public API and there's nothing stopping others from uploading chunks and not sending the header.

I think for now lets just not set dag size and wait for the cron job to set it.